### ISSUE TITLE
Adding Pre-Releases to F-Droid's Default Repo

### ISSUE THREAD

**Issue by Gitoffthelawn** [florisboard/florisboard#1896]
Could the pre-releases be added to F-Droid's default repo?

The pre-releases don't need to be tagged as "Recommended", so user's won't get auto-updated to pre-release versions (unless that is your preference).

F-Droid's default repo tends to lag a bit behind (not sure exactly why, or how to help), but besides that issue, it seems like adding FlorisBoard pre-releases to F-Droid's default repo could add a bunch more testers on a wide variety of devices.

Related: #462, #476

**Comment by tsiflimagas** [comment]
Beta releases are already available through the IzzyOnDroid repo.

**Comment by Gitoffthelawn** [comment]
> Beta releases are already available through the IzzyOnDroid repo.

Correct.  The proposal is to add them to the F-Droid default repo as well.  We might know about Izzy's repo, but compared to the main F-Droid repo, few people know about it.

**Comment by tsiflimagas** [comment]
I don't know how true is that. Most third party F-Droid clients include IzzyOnDroid by default, and I think many people prefer due to its faster updates delivery (myself included). But even so, I don't know if there's a need for more testers. There are already lots of people obtaining betas from IzzyOnDroid, GitHub and Playstore beta program and giving feedback. Also, F-Droid takes so much time to build and release updates, that there probably wouldn't be any point, as the next beta would have already been released. I guess these considerations were already made, else beta releases would already have been on F-Droid too. But, perhaps Patrick has a different take on this.

**Comment by Gitoffthelawn** [comment]
@tsiflimagas Do you happen to know why the default F-Droid takes so much time to build and release updates?  I've noticed that too.

**Comment by patrickgold** [comment]
Thanks for your proposal!

Like already mentioned, beta builds are distributed via the IzzyOnDroid F-Droid repository simply because it is just faster than the main F-Droid repo. Also If I tinker around with what's needed to build this project, I would find it quite annoying to adapt the build recipe for every beta release while I try around a few things, it is already enough if I have to do this for every stable release (provided something has changed in the way the build works).

> Most third party F-Droid clients include IzzyOnDroid by default

Most if not all major F-Droid clients have it in the repository selection page, however by default it is not selected. But I think if one wants to become a beta tester via F-Droid then he/she will find out quite simple how to enable the Izzy repository, and until now this did not pose a problem.

> There are already lots of people obtaining betas from IzzyOnDroid, GitHub and Playstore beta program and giving feedback

I have to agree with Kostas here too, there's already plenty of people on the beta track and everything works well, so I don't think there's a need just from the publishing perspective to put it on the main F-Droid repo.

> Do you happen to know why the default F-Droid takes so much time to build and release updates? I've noticed that too.

I am not Kostas but the reason why it takes so long for updates to appear on the main repo is because there's just a massive amount of apps that need constant re-indexing if there's a new version (that alone can take up to a day), if there is then they get queued for build. Build time typically varies from 5-10 minutes per app, but for some apps it takes much longer. After that the built APKs get grouped and at specific intervals I think they are taken to another server for signing with a digital key (required for all apps on Android to be signed). After that they get uploaded to the server and it still takes some time until the local F-Droid clients pick up the new update.

While sometimes it can be annoying that it takes so long for updates to appear, you have to give them a break here, because they just don't have the resources like Google has to be super fast with this process. Also they are doing an amazing work to ensure that everything corresponds to that exact source tar ball that they get for an app update, so you as a user get additional security and auditability of each app at the cost of build efficiency.

**Comment by Gitoffthelawn** [comment]
Thanks for your thoughtful response.

> Also If I tinker around with what's needed to build this project, I would find it quite annoying to adapt the build recipe for every beta release while I try around a few things, it is already enough if I have to do this for every stable release (provided something has changed in the way the build works).

How is the process for the IzzyOnDroid repo different from the main F-Droid repo?  It sounds like it is different.

One thing I especially like about the main F-Droid repo is that numerous people are involved, so any errors/violations of their process (of which they are transparent) will be unlikely, and noticed by more eyes.

> I am not Kostas but the reason why it takes so long for updates to appear on the main repo is because there's just a massive amount of apps that need constant re-indexing if there's a new version (that alone can take up to a day)...

I am not Kostas either (what if we were all Kostas? 🤣 ), but I do I happen to track all the new app versions on the main F-Droid repo, and most days, there are only a handful of them (there are certainly exceptions!).  I'm not sure where the bottlenecks are in their systems/procedures, but given that they are a very talented (and dedicated) bunch, I'm sure there are very good reasons.  I'm hoping donations help with buying additional hardware, if that's a bottleneck.

> Also they are doing an amazing work to ensure that everything corresponds to that exact source tar ball that they get for an app update, so you as a user get additional security and auditability of each app at the cost of build efficiency.

I agree.  They do amazing work.  F-Droid is a big deal, IMO, and they do a great job with it.

Does the IzzyOnDroid repo also perform the steps you outlined above?

**Comment by supermamie** [comment]
As far as I know, F-Droid builds all apps by themselves (in a new clean VM for each builds, with all tooling re-installed for each builds so there is no pollution from other tools needed for other builds).
Izzy's repo do NOT build anything, it uses the apk released by the authors.

So there us far less security about the content of an app in izzy's repo than tge official F-droid repo (but far more than google's repo a.k.a. Play Store)

This big differince is the main cause of the delay in F-Droid and why Izzy is so quick to have a new version.

This difference can also impact some users that knows about Izzy but do not want to use it because you can not have a proof of what is in the app.

(But as stated before, adding every beta to F-Droid means to work on the builds scripts far more, and no way to know if the beta will be released before it is obsolete).

**Comment by Gitoffthelawn** [comment]
> As far as I know, F-Droid builds all apps by themselves (in a new clean VM for each builds, with all tooling re-installed for each builds so there is no pollution from other tools needed for other builds).

This is my understanding as well.

> Izzy's repo do NOT build anything, it uses the apk released by the authors.

Ah, I did not know that.  Good to learn.  How does Izzy get the APKs?  From sites like GitHub, via email, or via another mechanism?  Seems like a lot to maintain, so I'm guessing it's automated, but that's just a guess.

> So there us far less security about the content of an app in izzy's repo than tge official F-droid repo (but far more than google's repo a.k.a. Play Store)

The first part of that sentence makes good sense to me.  The second part could use clarification: Why is Izzy's repo far more security-oriented than Google's Play Store?  Is it simply because all apps have open source code (which hopefully matches the APKs), or is there more to it?  (BTW, I don't know if all the apps in Izzy's repo are open source, but I seem to recall this may be the case.)

> This big differince is the main cause of the delay in F-Droid and why Izzy is so quick to have a new version.

Yes, that's a big (and important) difference!

> This difference can also impact some users that knows about Izzy but do not want to use it because you can not have a proof of what is in the app.

Yes, I prefer to know what's in the app, which is why I'm a supporter of F-Droid.  I think their build process is quite impressive.

> (But as stated before, adding every beta to F-Droid means to work on the builds scripts far more, and no way to know if the beta will be released before it is obsolete).

Given that an APK needs to be built either way, what overhead is F-Droid adding to the build scripts that isn't routinely done?  I ask because maybe other app devs are experiencing similar "pain points", and perhaps the process can be improved to help the F-Droid ecosystem.

@patrickgold 
I'm currently not using the beta versions of FlorisBoard, but I would like to help out by using them and providing constructive feedback to you.  If you decide to not make them available in the main F-Droid repo, I'll have several choices: build them myself; get them from Izzy's repo; get them from Google's Play Store; or don't use the betas.

As far as the pre-built FlorisBoard APKs go, are the Izzy-repo and Google Play Store APKs identical?  And if the betas were being built by F-Droid, would those APKs be the same as the ones on Izzy's repo and Google's Play Store?

Thanks for your help.  I think FlorisBoard is going to be the best keyboard for Android for many years to come.  I use it as my "daily driver", and I am continually impressed by it and where it's heading. 👍🏾

**Comment by supermamie** [comment]
Here are 2 links for more information :
For Izzy, mostly everything is explained here https://apt.izzysoft.de/fdroid/index/info
(how are the apk retreived, some part about security)

For F-Droid, the information is scattered but there was a recent blog post about the build process : https://f-droid.org/fr/2022/05/24/buildserver-overhaul-sponsored-by-calyx-institute.html

If you want to check what happens when florisboard is built, here is the log for 0.3.15 build : https://f-droid.org/repo/dev.patrickgold.florisboard_80.log.gz
TL;DR : total time 14min (including installation of Android NDK, Java, GCC and other tools)

For the part about Izzy being more secure than Google Play, here are the requirements to be included in Izzy's repo : https://gitlab.com/IzzyOnDroid/repo/#what-are-the-requirements-an-app-must-meet-to-be-included-with-the-repo (so yes, free and open source apps only). The rules of Google Play are someway obscure and tons of apps are full of ####, and some nice apps are kicked, see what happened to FairEmail recently for example : https://faircode.eu/faq.html , so it is really not clear :disappointed:

**Comment by patrickgold** [comment]
Sorry for my late reply, was a busy weekend.

Like I said earlier, atm there's no motivation for me to put the beta track on main F-Droid and I think this will not change until there's a good reason to do so. This has nothing to do though if main F-Droid is more or less secure than Izzy or other distribution means, but just that the current release model for beta works really well.

> I don't really see anything in the inclusion policy that would disqualify it, at a quick glance, so feel free to submit it to them and see where it goes.

That's true, why should it disqualify if the stable releases (which are published in main F-Droid) also adheres to all F-Droid rules? Basically the only difference between a beta and stable build is the package name and app icon/name.

> Either way, it's not up to an app's author to decide whether their app is added to F-Droid main or not. You'll have to submit it to them.

That's not true. There's the following line in the inclusion issue form:

`* [ ] The original app author has been notified (and does not oppose the inclusion).`

and I would oppose the inclusion of the beta track into main F-Droid because that would cause unnecessary confusion for those who have the IzzyOnDroid repo activated. Users may get outdated beta builds depending on which repo is prioritized in the user client, which is more than sub-optimal. I need early beta tester feedback to be able to more quickly fix bugs and issues I may have introduced.

As for all the other questions, @supermamie has provided an excellent summary of resources (thanks a lot!!) where one can read about the process in detail, so I won't really add more to this.

**Comment by patrickgold** [comment]
> What I meant by that is that it's not an app's author who decides on their own to add their releases to F-Droid; the request has to be made to F-Droid themselves. Of course, the author has the final say, but they aren't the ones who control what gets added. App authors don't have that kind of access ;p

Aaahhh you mean it that way. Then of course you are correct, F-Droid's maintainers have the last say if an app gets included after the author approved this. I thought you meant that anyone can just request the inclusion of an app while overriding the app author, thus my reply :)

**Comment by opusforlife2** [comment]
One important point I don't see mentioned here (maybe I missed it) is that F-Droid's default repo only keeps the last 3 builds for each app, other than some special cases (like the F-Droid app itself). This means that the moment the 3rd beta in a cycle is released on F-Droid, there would be no stable builds for new users to install. O_O

**Comment by supermamie** [comment]
> This means that the moment the 3rd beta in a cycle is released on F-Droid, there would be no stable builds for new users to install. O_O

This is just false.
The F-Droid repo can make difference between stable and unstable releases, and there will always be at least one stable release.
For example, there are a lot of unstable releases of K9-mail between stable releases, and there were always at least one stable release available. (now is a bad time to check as the last release is the stable one).

**Comment by patrickgold** [comment]
> One important point I don't see mentioned here (maybe I missed it) is that F-Droid's default repo only keeps the last 3 builds for each app, other than some special cases (like the F-Droid app itself). This means that the moment the 3rd beta in a cycle is released on F-Droid, there would be no stable builds for new users to install. O_O

This would not be the problem as the beta has a different package name and thus is technically a completely separate app for F-Droid, thus beta and stable builds are completely separated.

**Comment by Gitoffthelawn** [comment]
> Sorry for my late reply, was a busy weekend.

No problem at all.  Take your time... enjoy life!

Your comments have been very helpful... thank you.

I'm still on the main F-Droid repo (and will likely stick with it, but I haven't had time to read about Izzy's repo, so we'll see); I am eagerly awaiting the latest FlorisBoard release there.

I think the only small unanswered question is: As far as the pre-built FlorisBoard APKs go, are the Izzy-repo and Google Play Store APKs identical?

**Comment by yephny** [comment]
It should also be noted that the app's beta releases are naturally frequent and not intended to be the releases delivered to the Florisboard userbase. This leads me to think that having the beta versions in the main repository would be a waste of F-Droid's already limited resources.

**Comment by Gitoffthelawn** [comment]
> It should also be noted that the app's beta releases are naturally frequent and not intended to be the releases delivered to the Florisboard userbase. This leads me to think that having the beta versions in the main repository would be a waste of F-Droid's already limited resources.

Just FYI, unless an F-Droid release is tagged as `recommended`, F-Droid main repository users don't get updated, so there is minimal resource use.  OTOH, if you're referring to the build time, if F-Droid is simply in need of another build computer, let us know, as we can start a quick fundraiser for such a relatively inexpensive item.

**Comment by patrickgold** [comment]
> Just FYI, unless an F-Droid release is tagged as recommended, F-Droid main repository users don't get updated, so there is minimal resource use

As one of the most important things of beta builds is that they get fast to the end user (including auto-update if the user has this enabled), every beta release must be tagged as Recommended. Would technically not pose a problem as the beta has a separate package name, but still I like to use Izzy's repo more for beta builds as it does not have the build time delay.

**Comment by Gitoffthelawn** [comment]
@patrickgold Thanks Patrick.  0.3.16 hit the F-Droid main repo the other day, and it's working like a charm.  Great job!

There is much text above, so you probably missed this small question: As far as the pre-built FlorisBoard APKs go, are the Izzy-repo and Google Play Store APKs identical?

**Comment by patrickgold** [comment]
> As far as the pre-built FlorisBoard APKs go, are the Izzy-repo and Google Play Store APKs identical?

Depends on what you see as identical.

Binary they are not 100% identical, as Google Play does APK splitting, where they only deliver you the stuff that your device needs (so if you for instance have a 64-bit ARM CPI you only get the armv8a libs), also in some other areas like resolution-dependent images only the correct one gets included. That's Google's effort to reduce bandwidth usage and save data.

Functionally they are 100% identical, as they include the same logic and asset files (this can be confirmed by de-compiling the APK and check the generated source files). I also do not obfuscate any identifier/resource names, which should make the verification process easier.

**Comment by patrickgold** [comment]
The original intent of this issue is now obsolete, so I will close this issue as it is not needed anymore.
